File: rollup.js


$(function() {

  var $select = $('select');


  function addCities(element, index, array) {

    var groupFound = false,
        allGroups = $select.find('optgroup'),
        $group = {};

    if (allGroups.length > 0) {
      allGroups.each( function(index) {
        if ($(this).attr('label') === element.country) {
          groupFound = true;
          $group = $(this);
        }
      });
    }

    if (!groupFound) {
      $group = $('<optgroup label="' + element.country + '">');
      $select.append($group);
    }
    
    $group.append('<option value="' + element.city + '">' + element.city + '</option>');
  }


  $.getJSON( "cities.json", function( data ) {
    data.forEach(addCities);
  });
});

File: rollup.rb
# encoding: utf-8

require 'json'

def load_from_file( filename )
  File.open( filename, "r" ) do |f|
    JSON.load( f )
  end
end

grouped_by_cities = load_from_file('cities.json')
grouped_by_countries = Hash.new


grouped_by_cities.each do |city|

  # create the country if not yet in hash
  grouped_by_countries.store(city["country"], []) unless grouped_by_countries.has_key?(city["country"])

  # group cities into each country
  grouped_by_countries[city["country"]] << city["city"]
end

puts grouped_by_countries

File.open("countries.json","w") do |f|
  f.write(grouped_by_countries.to_json)
end